expo/expo
# [android] Use stable manifest ID where applicable

# Why

This PR audits all `RawManifest.getID` callsites and ensures that `scopeKey` (preferred, used in all cases other than expo push registration and auth session) or `originalFullName` are used instead of `id` when possible and applicable. `scopeKey` and `originalFullName` are stable through project transfers so they should be used for client data scoping and legacy API scoping respectively, but should not be used for things like manifest loading or manifest caching at this point.

# How

This PR accomplishes this by performing the following steps.
1. Split `RawManifest.getID` into three new methods and then use the compiler to tell me what needs attention:
    - `getLegacyID` - equivalent to `getID` and used for non-scoped purposes. For example, when storing manifests in the local DB we want to store them by their `id` field still, not the original `id` of the project.
    - `getStableLegacyID` - `originalFullName` with `id` as a fallback. Used for all scoping that previously used `id` for scoping and requires first or third party API calls (amplitude and push notifications for example).
    - `getScopeKey` - `scopeKey` with `getStableLegacyID` as fallback. Used for all scoping on the client that doesn't require external API calls (shared preferences, expo-task-manager scoping, etc...)
2. Replace the `ExperienceId` class (wrapper around the experience id string) with `ExperienceKey`, which is a wrapper around scope key from the manifest. This class is a kotlin data class, so two instances will be equal if their values are equal (same semantics as `ExperienceId` but without the constructor memoization since it seemed unnecessary).
3. Use `ExperienceKey` in as many internal APIs as possible and then only at the leafs where an actual value is needed (to key into shared preferences for example) do we extract the scopeKey.

# Concerns

- At this size, this PR is pretty much guaranteed to have mistakes. This is compounded by the need to manually update 4 versions of the code for every versioned code change. Unfortunately, splitting this up proved to be difficult due to the nature of the change not being able to be done incrementally to retain the safety of the compiler (removing types entirely to find callsites that need attention). 
- Testing this thoroughly is fairly unfeasible as it requires building an app with all the features touched here, transferring the app, running again, creating a new build in the new account, re-running the app, and then pushing an OTA update on the new account for the app. Throughout this four step process we'd need to check all of the following: push notifications, error recovery (we'd need to trigger an error), permissions, sensors, schedulers (scheduled notifications), notification groups, pedometer, facebook, accelerometer, gyroscope, gravity sensor (?), linear acceleration, magnetometer, rotation, tasks (expo-task-manager), and probably some that I've missed.
- Instead of manually testing, here's what I propose for reviewers (I've already run through this, but for code review this is a high leverage review strategy):
    1. Assume compiler is fairly good at figuring out what places needed attention due to method renames (removed `getID` and `ExperienceId` for example).
    2. Check out the branch locally. Use android studio to "Find Usages" of the three methods in `RawManifest` and the corresponding three methods in `ExperienceKey`. Ensure that scopeKey, stableLegacyID, and legacyId are used where expected.

# Test Plan

- Load an experience that has been transferred between accounts, ensure the fields are set to the correct thing in `ExperienceActivity` using a breakpoint. Send a notification to a transferred experience, see that it works.
- The 4 step plan specified in the concerns section.
- Load native component list, use all APIs, see that nothing is broken at first glance, especially the notifications test.
- @tcdavis and I will do a bunch of dogfooding in the future (of all the modules mentioned above) to check all the modules that are scoped in a standalone context, but for now it is tough.
